### PR TITLE
Standardize join internals around DataFrame

### PR DESCRIPTION
**Pull request by vyasr**
This PR un-deprecates the `BaseIndex.join` method, which was erroneously deprecated, but reimplements it by converting the index objects to DataFrames under the hood. This reimplementation allows us to simplify much of the internals around merging, providing a single main code path implemented only for DataFrames. The behavior of `cudf.merge(Series, DataFrame)` is preserved via an explicit upcast as well. These changes allow further simplification of the (currently extremely complex) internals of merging, which hopefully will eventually allow us to extract a fast and simple merge function for internal use from the complexities of the public DataFrame.merge API. This change also removes any vestigial accesses to the Frame._index to enable us to remove that.

### CODE REVIEW

**Comment by vyasr** [comment]
rerun tests

**Comment by vyasr** [comment]
rerun tests

File: python/cudf/cudf/core/dataframe.py [hunk 1]
```diff
@@ -3559,9 +3560,39 @@ def merge(
...
+            how = "left"
+            lhs, rhs = right, self
+            left_on, right_on = right_on, left_on
+            left_index, right_index = right_index, left_index
+            suffixes = (suffixes[1], suffixes[0])
+        elif how in {"leftsemi", "leftanti"}:
+            merge_cls = MergeSemi
+
+        # TODO: the two isinstance checks below indicates that `_merge` should
+        # not be defined in `Frame`, but in `IndexedFrame`.
```
Comment (vyasr) [inline python/cudf/cudf/core/dataframe.py:3592]:
```suggestion
```

**Review by isVoid — APPROVED** [review]

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**